primitives/traits/src/alias.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+#[doc(hidden)]
+pub mod __reexports {
+    pub use {
+        frame_support::{CloneNoBound, EqNoBound, PartialEqNoBound, RuntimeDebugNoBound},
+        scale_info::TypeInfo,
+        serde::{de::DeserializeOwned, Deserialize, Serialize},
+        sp_core::{Decode, Encode, RuntimeDebug},
+    };
+}
+
+pub use macro_rules_attribute::apply;
+
+#[macro_export]
+macro_rules! derive_storage_traits {
+    ( $( $tt:tt )* ) => {
+        #[derive(
+            $crate::alias::__reexports::RuntimeDebug,
+            ::core::cmp::PartialEq,
+            ::core::cmp::Eq,
+            ::core::clone::Clone,
+            $crate::alias::__reexports::Encode,
+            $crate::alias::__reexports::Decode,
+            $crate::alias::__reexports::TypeInfo,
+            $crate::alias::__reexports::Serialize,
+            $crate::alias::__reexports::Deserialize,
+        )]
+        $($tt)*
+    }
+}
+
+/// Derives traits related to SCALE encoding and serde.
+#[macro_export]
+macro_rules! derive_scale_and_serde {
+    ( $( $tt:tt )* ) => {
+        #[derive(
+            $crate::alias::__reexports::Encode,
+            $crate::alias::__reexports::Decode,
+            $crate::alias::__reexports::TypeInfo,
+            $crate::alias::__reexports::Serialize,
+            $crate::alias::__reexports::Deserialize,
+        )]
+        $($tt)*
+    }
+}
+
+// This currently doesn't work due to a quirk in RuntimeDebugNoBound, PartialEqNoBound
+// and CloneNoBound, as there seem to be something breaking macro hygiene. This is not an
+// issue when using the derive directly, but doesn't compile when adding it through our macro.
```

Comment:
What problems? I tried to use this macro in `pallets/data-preservers/src/types.rs` and it compiles fine:

```diff
-#[derive(RuntimeDebugNoBound, PartialEqNoBound, EqNoBound, CloneNoBound)]
-#[apply(derive_scale_and_serde)]
+#[apply(derive_storage_traits_no_bounds)]
 #[scale_info(skip_type_params(T))]
 pub enum ParaIdsFilter<T: Config> {
```

Reply:
Really? I spent quite some time pulling my hair on this one, and it didn't worked even after doing a `cargo clean`

Reply:
were we able to solve this @nanocryk ?

Reply:
It turns out the macro is implemented wrong and it won't work in an alias, except in some cases such as the `ParaIdsFilter` because it is an enum, but for structs the `*NoBound` macros cannot be aliased

Reply:
how important it is? do we use it?

Reply:
@nanocryk maybe we can create an issue to PR polkadot about the macro not being correctly implemented?